### DISCUSSION TITLE
Question on how filereader works

### DISCUSSION THREAD

**Discussion by rutledj** [nextgenhealthcare/connect#5049] [Q&A]
I'm curious how the filereader handles this situation.  An sftp reader connects to remote server (runs only once a day) and starts reading in files. Let's say there are 5k files there.  Transfer starts.  Does the filereader constantly get a new list of files periodically (during the same initial read).  I'm wondering if files are being put in the folder while the fiereader is reading, does it keep going until all files (including new ones) are read?

**Comment by jonbartels** [comment]
You might be able to interpret that from the source code. The SFTP magic is in https://github.com/nextgenhealthcare/connect/blob/e66d1122dbeebfcf5a8939beb93d720b2d5071fb/server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java then that connection is used in https://github.com/nextgenhealthcare/connect/blob/e66d1122dbeebfcf5a8939beb93d720b2d5071fb/server/src/com/mirth/connect/connectors/file/FileReceiver.java#L180 .

There may even be some log statements that show the exact FTP commands issued and when.

**Comment by rutledj** [comment]
Thanks. Not exactly sure where it is building the list of files in that code (when not using recursive directory).

'''
else {
                processFiles(listFiles(readDir), pollId, pollSequenceId, true);
'''

listFiles(readDir), how is that populated?

**Reply by ab-20-m** [reply]
https://github.com/nextgenhealthcare/connect/blob/e66d1122dbeebfcf5a8939beb93d720b2d5071fb/server/src/com/mirth/connect/connectors/file/filesystems/SftpConnection.java#L203-L221

I've never dug into the depths of how SFTP, much less ls works internally. So I'm not sure what the behavior would be but it could be file system dependent. I wouldn't make any assumptions about ls being transactional/atomic with respect to listed files.

**Reply by jonbartels** [reply]
FWIW I am curious how this works too. It is a good question. I think @ab-20-m is correct - This is more about how the SFTP server and it's filesystem handles concurrent operations.

As a practical matter, do you just need to set the "file age" setting in MC so that a file which is not fully written is not read by your polling operation? That would prevent a partially-written file from being read by the SFTP `ls` operation.

**Comment by rutledj** [comment]
I was just trying to understand if the channel polls one time (at it designated interval) and only processes that initial grab of files (and any other added files after that won't get pulled) or if it somehow constantly polls until all files are moved (from the beginning interval).
